aten/src/ATen/NestedTensorImpl.cpp

```diff
@@ -209,5 +241,44 @@ const char* NestedTensorImpl::tensorimpl_type_name() const {
   return "NestedTensorImpl";
 }
 
+
+template <typename VariableVersion>
+c10::intrusive_ptr<TensorImpl> NestedTensorImpl::shallow_copy_and_detach_core(
+    VariableVersion&& version_counter,
+    bool allow_tensor_metadata_change) const {
+  if (key_set_.has(DispatchKey::Python) &&
+      !c10::impl::tls_is_dispatch_key_excluded(DispatchKey::Python)) {
+    auto r = pyobj_interpreter_.load(std::memory_order_acquire)->detach(this);
+    if (r) {
+      r->set_version_counter(std::forward<VariableVersion>(version_counter));
+      r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
+      return r;
+    }
+    // otherwise just copy the TensorImpl and not the PyObject.  Since
+    // the interpreter is dead no one can call us out on it
+  }
+  auto impl = c10::make_intrusive<NestedTensorImpl>(this -> buffer_, this -> nested_size_tensor_);
```

Comment:
Since softmax_backward uses SavedVariable() `shallow_copy_and_detach` gets called. The base class`TensorImpl` calls `impl->refresh_numel();` which errors for nested_tensors. Thefore we override this method and instead construct a `NestedTensorImpl` and remove the `refresh_numel` and `refresh_contiguous` call.

This PR also introduces a NestedTensorImpl constructor which accepts a key_set and sets the `ImplType` to `c10::TensorImpl::VIEW` .  I don't think though that the above constructor needs to be changed. `copy_tensor_metadata` will copy over the key_set and all relevant metadata that gets "lost" in the above constructor. The only difference is that the `ImplType` will not be set.  However, from looking at the base class shallow_copy method it does not appear that this is required. 

@albanD @soulitzer  feel free to correct me

Reply:
Don't you need to copy the NestedTensorImpl specific fields here? copy_tensor_metadata is only copying the metadata from the base TensorImpl no?

Reply:
yeah good call out, in the contiguous case the constructor used would create the required fields but should be directly copying all the nested specific meta data.